// webwaspjs/dist/index.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.js
var index_exports = {};
__export(index_exports, {
  Aggregation: () => Aggregation,
  Collider: () => Collider,
  Connection: () => Connection,
  Part: () => Part,
  Plane: () => Plane,
  PlaneToPlane: () => PlaneToPlane,
  Rule: () => Rule,
  Visualizer: () => Visualizer,
  checkMeshesIntersection: () => checkMeshesIntersection,
  createAggregationFromData: () => createAggregationFromData,
  generateRules: () => generateRules,
  getAggregationCatalogParts: () => getAggregationCatalogParts,
  getOpenConnectionsForPart: () => getOpenConnectionsForPart,
  getValidPlacementsAtConnection: () => getValidPlacementsAtConnection,
  getValidPlacementsAtParent: () => getValidPlacementsAtParent,
  meshFromData: () => meshFromData,
  meshToData: () => meshToData,
  setActivePartTypes: () => setActivePartTypes,
  transformFromData: () => transformFromData,
  transformToData: () => transformToData,
  types: () => types
});
module.exports = __toCommonJS(index_exports);

// src/core/utilities.js
var THREE = __toESM(require("three"), 1);

// src/core/rule.js
var Rule = class _Rule {
  /**
   * @param {string} partA
   * @param {number} connectionA
   * @param {string} partB
   * @param {number} connectionB
   */
  constructor(partA, connectionA, partB, connectionB) {
    this.partA = partA;
    this.connectionA = connectionA;
    this.partB = partB;
    this.connectionB = connectionB;
    this.active = true;
  }
  /**
   * Build a rule from serialized data.
   * @param {import('../types').RuleData} data
   * @returns {Rule}
   */
  static fromData(data) {
    const rule = new _Rule(data.part1, data.conn1, data.part2, data.conn2);
    rule.active = data.active;
    return rule;
  }
  /**
   * Serialize the rule.
   * @returns {import('../types').RuleData}
   */
  toData() {
    return {
      part1: this.partA,
      conn1: this.connectionA,
      part2: this.partB,
      conn2: this.connectionB,
      active: this.active
    };
  }
};

// src/core/utilities.js
var import_three_mesh_bvh = require("three-mesh-bvh");
if (THREE.Mesh.prototype.raycast !== import_three_mesh_bvh.acceleratedRaycast) {
  THREE.Mesh.prototype.raycast = import_three_mesh_bvh.acceleratedRaycast;
}
function checkMeshesIntersection(mesh1, mesh2) {
  if (!mesh1.geometry.boundsTree) {
    mesh1.geometry.boundsTree = new import_three_mesh_bvh.MeshBVH(mesh1.geometry);
  }
  if (!mesh2.geometry.boundsTree) {
    mesh2.geometry.boundsTree = new import_three_mesh_bvh.MeshBVH(mesh2.geometry);
  }
  const transformMatrix = new THREE.Matrix4();
  transformMatrix.copy(mesh2.matrixWorld).invert().multiply(mesh1.matrixWorld);
  const intersects = mesh2.geometry.boundsTree.intersectsGeometry(mesh1.geometry, transformMatrix);
  return intersects;
}
function meshToData(mesh) {
  const geom = mesh.geometry;
  const data = { vertices: [], faces: [] };
  if (geom.isBufferGeometry) {
    const pos = geom.getAttribute("position");
    const index = geom.getIndex();
    for (let i = 0; i < pos.count; i++) {
      data.vertices.push([pos.getX(i), pos.getY(i), pos.getZ(i)]);
    }
    if (index) {
      for (let i = 0; i < index.count; i += 3) {
        data.faces.push([index.getX(i), index.getX(i + 1), index.getX(i + 2)]);
      }
    } else {
      for (let i = 0; i < pos.count; i += 3) {
        data.faces.push([i, i + 1, i + 2]);
      }
    }
    return data;
  }
  geom.vertices.forEach((v) => data.vertices.push([v.x, v.y, v.z]));
  geom.faces.forEach((f) => {
    const face = [f.a, f.b, f.c];
    if (f.d !== void 0) face.push(f.d);
    data.faces.push(face);
  });
  return data;
}
function meshFromData(data) {
  const geometry = new THREE.BufferGeometry();
  const vertices = [];
  const indices = [];
  data.vertices.forEach((v) => {
    vertices.push(v[0], v[1], v[2]);
  });
  data.faces.forEach((f) => {
    if (f.length === 3) {
      indices.push(f[0], f[1], f[2]);
    } else if (f.length === 4) {
      indices.push(f[0], f[1], f[2], f[0], f[2], f[3]);
    }
  });
  geometry.setAttribute("position", new THREE.Float32BufferAttribute(vertices, 3));
  geometry.setIndex(indices);
  geometry.computeVertexNormals();
  const material = new THREE.MeshStandardMaterial({ color: 16777215, roughness: 0.5 });
  return new THREE.Mesh(geometry, material);
}
function transformToData(trans) {
  const e = trans.elements;
  return {
    M00: e[0],
    M01: e[4],
    M02: e[8],
    M03: e[12],
    M10: e[1],
    M11: e[5],
    M12: e[9],
    M13: e[13],
    M20: e[2],
    M21: e[6],
    M22: e[10],
    M23: e[14],
    M30: e[3],
    M31: e[7],
    M32: e[11],
    M33: e[15]
  };
}
function transformFromData(data) {
  const m = [
    parseFloat(data.M00),
    parseFloat(data.M01),
    parseFloat(data.M02),
    parseFloat(data.M03),
    parseFloat(data.M10),
    parseFloat(data.M11),
    parseFloat(data.M12),
    parseFloat(data.M13),
    parseFloat(data.M20),
    parseFloat(data.M21),
    parseFloat(data.M22),
    parseFloat(data.M23),
    parseFloat(data.M30),
    parseFloat(data.M31),
    parseFloat(data.M32),
    parseFloat(data.M33)
  ];
  const matrix = new THREE.Matrix4();
  matrix.set(
    m[0],
    m[1],
    m[2],
    m[3],
    m[4],
    m[5],
    m[6],
    m[7],
    m[8],
    m[9],
    m[10],
    m[11],
    m[12],
    m[13],
    m[14],
    m[15]
  );
  return matrix;
}
function PlaneToPlane(sourcePlane, targetPlane) {
  const sourceMatrix = new THREE.Matrix4();
  sourceMatrix.makeBasis(
    sourcePlane.xaxis.clone().normalize(),
    sourcePlane.yaxis.clone().normalize(),
    sourcePlane.zaxis.clone().normalize()
  );
  sourceMatrix.setPosition(sourcePlane.origin.clone());
  const targetMatrix = new THREE.Matrix4();
  targetMatrix.makeBasis(
    targetPlane.xaxis.clone().normalize(),
    targetPlane.yaxis.clone().normalize(),
    targetPlane.zaxis.clone().normalize()
  );
  targetMatrix.setPosition(targetPlane.origin.clone());
  const transformMatrix = new THREE.Matrix4();
  transformMatrix.copy(targetMatrix).multiply(sourceMatrix.invert());
  return transformMatrix;
}
function generateRules(parts, selfPart = true, selfConnection = false, useTypes = false, grammar = []) {
  const rules = [];
  if (grammar.length === 0) {
    for (const part of parts) {
      for (const conn of part.connections) {
        for (const otherPart of parts) {
          let skipPart = false;
          if (!selfPart && part.name === otherPart.name) {
            skipPart = true;
          }
          if (!skipPart) {
            for (const otherConn of otherPart.connections) {
              let skipConn = false;
              if (!selfConnection && part.name === otherPart.name && conn.id === otherConn.id) {
                skipConn = true;
              }
              if (!skipConn) {
                if (useTypes) {
                  if (conn.type === otherConn.type) {
                    rules.push(new Rule(part.name, conn.id, otherPart.name, otherConn.id));
                  }
                } else {
                  rules.push(new Rule(part.name, conn.id, otherPart.name, otherConn.id));
                }
              }
            }
          }
        }
      }
    }
  } else {
    for (const gr_rule of grammar) {
      const [startType, endType] = gr_rule.split(">");
      for (const part of parts) {
        for (const conn of part.connections) {
          if (conn.type === startType) {
            for (const otherPart of parts) {
              let skipPart = false;
              if (!selfPart && part.name === otherPart.name) {
                skipPart = true;
              }
              if (!skipPart) {
                for (const otherConn of otherPart.connections) {
                  if (otherConn.type === endType) {
                    let skipConn = false;
                    if (!selfConnection && conn.id === otherConn.id) {
                      skipConn = true;
                    }
                    if (!skipConn) {
                      rules.push(new Rule(part.name, conn.id, otherPart.name, otherConn.id));
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  }
  return rules;
}

// src/core/plane.js
var THREE2 = __toESM(require("three"), 1);
var Plane = class _Plane {
  /**
   * @param {THREE.Vector3} origin
   * @param {THREE.Vector3} xaxis
   * @param {THREE.Vector3} zaxis
   */
  constructor(origin, xaxis, zaxis) {
    this.origin = origin.clone();
    this.xaxis = xaxis.clone().normalize();
    this.zaxis = zaxis.clone().normalize();
    this.yaxis = new THREE2.Vector3().crossVectors(this.xaxis, this.zaxis).normalize();
  }
  clone() {
    return new _Plane(this.origin.clone(), this.xaxis.clone(), this.zaxis.clone());
  }
  applyMatrix4onPlane(matrix) {
    let newPlane = this.clone();
    newPlane.origin.applyMatrix4(matrix);
    const rotationMatrix = new THREE2.Matrix4().extractRotation(matrix);
    newPlane.xaxis.applyMatrix4(rotationMatrix).normalize();
    newPlane.yaxis.applyMatrix4(rotationMatrix).normalize();
    newPlane.zaxis.applyMatrix4(rotationMatrix).normalize();
    return newPlane;
  }
  /**
   * Serialize plane axes.
   */
  toData() {
    return {
      origin: [this.origin.x, this.origin.y, this.origin.z],
      xaxis: [this.xaxis.x, this.xaxis.y, this.xaxis.z],
      yaxis: [this.yaxis.x, this.yaxis.y, this.yaxis.z],
      zaxis: [this.zaxis.x, this.zaxis.y, this.zaxis.z]
    };
  }
  /**
   * Build a plane from serialized data (supports legacy zaxis via yaxis fallback).
   */
  static fromData(data) {
    const origin = new THREE2.Vector3(data.origin[0], data.origin[1], data.origin[2]);
    const xaxis = new THREE2.Vector3(data.xaxis[0], data.xaxis[1], data.xaxis[2]);
    const zAxisSource = data.zaxis ?? data.yaxis;
    const zaxis = new THREE2.Vector3(zAxisSource[0], zAxisSource[1], zAxisSource[2]);
    return new _Plane(origin, xaxis, zaxis);
  }
};

// src/core/connection.js
var Connection = class _Connection {
  /**
   * @param {Plane} _plane
   * @param {string} _type
   * @param {string} _part
   * @param {number} _id
   */
  constructor(_plane, _type, _part, _id) {
    this.pln = _plane.clone();
    const tempPlane = this.pln.clone();
    this.flip_pln = new Plane(tempPlane.origin, tempPlane.xaxis, tempPlane.zaxis.clone().multiplyScalar(-1));
    this.type = _type;
    this.part = _part;
    this.id = _id;
    this.rules_table = [];
    this.active_rules = [];
  }
  /**
   * Deserialize a connection.
   */
  static fromData(data) {
    let c_pln = Plane.fromData(data["plane"]);
    return new _Connection(c_pln, data["type"], data["part"], parseInt(data["id"]));
  }
  /**
   * Serialize the connection.
   */
  toData() {
    return {
      plane: this.pln.toData(),
      type: this.type,
      part: this.part,
      id: this.id
    };
  }
  transform(trans) {
    let newPlane = this.pln.applyMatrix4onPlane(trans);
    return new _Connection(newPlane, this.type, this.part, this.id);
  }
  copy() {
    let pln_copy = this.pln.clone();
    return new _Connection(pln_copy, this.type, this.part, this.id);
  }
  generateRulesTable(rules) {
    let count = 0;
    this.rules_table = [];
    this.active_rules = [];
    rules.forEach((rule) => {
      if (rule.partA === this.part && rule.connectionA === this.id) {
        this.rules_table.push(rule);
        this.active_rules.push(count);
        count += 1;
      }
    });
  }
};

// src/core/collider.js
var THREE3 = __toESM(require("three"), 1);
function meshesIntersectByBounds(meshA, meshB) {
  if (!meshA || !meshB) return false;
  meshA.updateMatrixWorld?.(true);
  meshB.updateMatrixWorld?.(true);
  const boxA = new THREE3.Box3().setFromObject(meshA);
  const boxB = new THREE3.Box3().setFromObject(meshB);
  return boxA.intersectsBox(boxB);
}
var Collider = class _Collider {
  /**
   * @param {Array<THREE.Mesh>} geometry
   * @param {boolean} multiple
   * @param {boolean} checkAll
   * @param {Connection[]} connections
   * @param {number[]} validConnections
   */
  constructor(geometry, multiple = false, checkAll = false, connections = [], validConnections = []) {
    this.geometry = geometry;
    this.multiple = multiple;
    this.checkAll = checkAll;
    this.connections = connections;
    this.facesCount = 0;
    if (this.geometry && Array.isArray(this.geometry)) {
      for (let geo of this.geometry) {
        if (geo && geo.faces) {
          this.facesCount += geo.faces.length;
        }
      }
    }
    this.validConnections = validConnections;
    this.setConnections = false;
    if (this.connections.length === this.geometry.length && this.multiple) {
      this.setConnections = true;
    }
  }
  /**
   * Deserialize collider data.
   */
  static fromData(data) {
    const geometry = data.geometry.map(meshFromData);
    const multiple = data.multiple;
    const checkAll = data.check_all;
    const connections = data.connections.map(Connection.fromData);
    const validConnections = data.valid_connections.map((vc) => parseInt(vc, 10));
    return new _Collider(geometry, multiple, checkAll, connections, validConnections);
  }
  /**
   * Serialize collider data.
   */
  toData() {
    return {
      geometry: this.geometry.map(meshToData),
      multiple: this.multiple,
      check_all: this.checkAll,
      connections: this.connections.map((conn) => conn.toData()),
      valid_connections: this.validConnections
    };
  }
  transform(trans, transformConnections = false, maintainValid = false) {
    const geometryTrans = this.geometry.map((geo) => {
      const geoTrans = geo.clone();
      geoTrans.applyMatrix4(trans);
      return geoTrans;
    });
    let connectionsTrans = [];
    if (transformConnections) {
      connectionsTrans = this.connections.map((conn) => conn.transform(trans));
    }
    if (maintainValid) {
      return new _Collider(geometryTrans, this.multiple, this.checkAll, connectionsTrans, [...this.validConnections]);
    } else {
      return new _Collider(geometryTrans, this.multiple, this.checkAll, connectionsTrans);
    }
  }
  copy() {
    const geometryCopy = this.geometry.map((geo) => geo.clone());
    const connectionsCopy = this.connections.map((conn) => conn.copy());
    const validConnectionsCopy = [...this.validConnections];
    return new _Collider(geometryCopy, this.multiple, this.checkAll, connectionsCopy, validConnectionsCopy);
  }
  checkCollisionsWithParts(parts) {
    if (this.multiple) {
      let validColliders = [];
      this.validConnections = [];
      let count = 0;
      for (let geo of this.geometry) {
        let validColl = true;
        for (let part of parts) {
          for (let otherGeo of part.collider.geometry) {
            if (meshesIntersectByBounds(geo, otherGeo)) {
              validColl = false;
              break;
            }
          }
          if (!validColl) break;
        }
        validColliders.push(validColl);
        if (this.setConnections && validColl) {
          this.validConnections.push(count);
        }
        if (validColl && !this.checkAll) break;
        count++;
      }
      return !validColliders.includes(true);
    } else {
      for (let geo of this.geometry) {
        for (let part of parts) {
          for (let otherGeo of part.collider.geometry) {
            if (meshesIntersectByBounds(geo, otherGeo)) {
              return true;
            }
          }
        }
      }
      return false;
    }
  }
  checkCollisionsById(parts, ids) {
    if (this.multiple) {
      let validColliders = [];
      for (let geo of this.geometry) {
        let validColl = true;
        for (let id of ids) {
          for (let otherGeo of parts[id].collider.geometry) {
            if (meshesIntersectByBounds(geo, otherGeo)) {
              validColl = false;
              break;
            }
          }
          if (!validColl) break;
        }
        validColliders.push(validColl);
        if (validColl && !this.checkAll) break;
      }
      return !validColliders.includes(true);
    } else {
      for (let geo of this.geometry) {
        for (let id of ids) {
          for (let otherGeo of parts[id].collider.geometry) {
            if (meshesIntersectByBounds(geo, otherGeo)) {
              return true;
            }
          }
        }
      }
      return false;
    }
  }
  checkIntersectionWithLine(line) {
    for (let geo of this.geometry) {
      if (geo.intersectsLine(line)) {
        return true;
      }
    }
    return false;
  }
  checkGlobalConstraints(_constraint) {
    return false;
  }
};

// src/core/part.js
var THREE4 = __toESM(require("three"), 1);
function parsePossiblyNumericId(value) {
  if (value === null || value === void 0 || value === "") {
    return null;
  }
  if (typeof value === "number" && !Number.isNaN(value)) {
    return value;
  }
  if (typeof value === "string") {
    const parsed = Number.parseInt(value, 10);
    return Number.isNaN(parsed) ? value : parsed;
  }
  return value;
}
var Part = class _Part {
  /**
   * @param {string} name
   * @param {THREE.Mesh} geometry
   * @param {Connection[]} connections
   * @param {Collider} collider
   * @param {Array} attributes
   * @param {number|null} dim
   * @param {number|string|null} id
   * @param {unknown} field
   */
  constructor(name, geometry, connections, collider = null, attributes, dim = 1, id = null, field = null) {
    this.name = name;
    this.id = id;
    this.geo = geometry;
    this.field = field;
    this.connections = [];
    this.active_connections = [];
    let count = 0;
    for (let conn of connections) {
      conn.id = count;
      this.active_connections.push(count);
      this.connections.push(conn);
      count++;
    }
    this.transformation = new THREE4.Matrix4().identity();
    const { center, dim: computedDim } = computeCenterAndDim(geometry, collider);
    this.center = center;
    this.collider = collider;
    this.dim = dim !== null ? dim : computedDim;
    this.parent = null;
    this.conn_on_parent = null;
    this.conn_to_parent = null;
    this.children = [];
    this.attributes = [];
    this.is_constrained = false;
  }
  // create class from data dictionary
  static fromData(data) {
    let p_name = data["name"];
    let p_geometry = meshFromData(data["geometry"]);
    let p_connections = data["connections"].map((c_data) => Connection.fromData(c_data));
    let p_collider = Collider.fromData(data["collider"]);
    let p_attributes = [];
    let p_dim = parseFloat(data["dim"]);
    let p_id = parsePossiblyNumericId(data["id"]);
    let p_field = data["field"];
    let part = new _Part(p_name, p_geometry, p_connections, p_collider, p_attributes, p_dim, p_id, p_field);
    part.transformation = new THREE4.Matrix4().identity();
    part.parent = data["parent"];
    part.conn_on_parent = data["conn_on_parent"];
    part.conn_to_parent = data["conn_to_parent"];
    for (let child_data of data["children"] || []) {
      part.children.push(parsePossiblyNumericId(child_data));
    }
    return part;
  }
  // return the data dictionary representing the part
  toData() {
    return {
      "class_type": "Part",
      "name": this.name,
      "id": this.id,
      "geometry": meshToData(this.geo),
      "field": this.field,
      "connections": this.connections.map((conn) => conn.toData()),
      "active_connections": this.active_connections,
      "collider": this.collider.toData(),
      "transform": transformToData(this.transformation),
      "dim": this.dim,
      "parent": this.parent,
      "conn_on_parent": this.conn_on_parent,
      "conn_to_parent": this.conn_to_parent,
      "children": this.children
    };
  }
  // reset the part and connections according to new provided aggregation rules
  resetPart(rules) {
    let count = 0;
    this.active_connections = [];
    for (let conn of this.connections) {
      conn.generateRulesTable(rules);
      this.active_connections.push(count);
      count++;
    }
  }
  // return a dictionary containing all part data
  returnPartData() {
    return {
      "name": this.name,
      "id": this.id,
      "geo": this.geo,
      "connections": this.connections,
      "transform": this.transformation,
      "collider": this.collider,
      "center": this.center,
      "parent": this.parent,
      "children": this.children,
      "attributes": this.attributes
    };
  }
  // return a transformed copy of the part
  transform(trans, _transform_sub_parts = false, maintain_parenting = false) {
    let geo_trans = this.geo.clone();
    geo_trans.applyMatrix4(trans);
    let collider_trans = this.collider.transform(trans);
    let connections_trans = [];
    for (let conn of this.connections) {
      let conn_trans = conn.transform(trans);
      connections_trans.push(conn_trans);
    }
    let attributes_trans = [];
    if (this.attributes.length > 0) {
      attributes_trans = this.attributes.map((attr) => attr.transform(trans));
    }
    let part_trans = new _Part(this.name, geo_trans, connections_trans, collider_trans, attributes_trans, this.dim, this.id, this.field);
    part_trans.transformation.copy(this.transformation).multiply(trans);
    if (maintain_parenting) {
      part_trans.parent = this.parent;
      part_trans.conn_on_parent = this.conn_on_parent;
      part_trans.conn_to_parent = this.conn_to_parent;
      part_trans.children = [...this.children];
    }
    return part_trans;
  }
  // return a copy of the part
  copy(maintain_parenting = false) {
    let geo_copy = this.geo.clone();
    let collider_copy = this.collider.copy();
    let connections_copy = this.connections.map((conn) => conn.copy());
    let attributes_copy = [];
    if (this.attributes.length > 0) {
      attributes_copy = this.attributes.map((attr) => attr.copy());
    }
    let part_copy = new _Part(this.name, geo_copy, connections_copy, collider_copy, attributes_copy, this.dim, this.id, this.field);
    part_copy.transformation.copy(this.transformation);
    if (maintain_parenting) {
      part_copy.parent = this.parent;
      part_copy.conn_on_parent = this.conn_on_parent;
      part_copy.conn_to_parent = this.conn_to_parent;
      part_copy.children = [...this.children];
    }
    return part_copy;
  }
  // return transformed center point of the part
  transformCenter(trans) {
    const centerTrans = this.center.clone();
    centerTrans.applyMatrix4(trans);
    return centerTrans;
  }
  // return transformed collider
  transformCollider(trans) {
    return this.collider.transform(trans);
  }
  assignId(id) {
    this.id = id;
  }
};
function computeCenterAndDim(mesh, collider) {
  const defaultResult = { center: new THREE4.Vector3(), dim: 1 };
  if (!mesh || !mesh.geometry) {
    return defaultResult;
  }
  if (typeof mesh.updateMatrix === "function") {
    mesh.updateMatrix();
  }
  const geom = mesh.geometry;
  if (!geom.boundingBox) {
    geom.computeBoundingBox();
  }
  if (!geom.boundingSphere) {
    geom.computeBoundingSphere();
  }
  const center = geom.boundingBox ? geom.boundingBox.getCenter(new THREE4.Vector3()) : defaultResult.center.clone();
  center.applyMatrix4(mesh.matrix);
  const radius = geom.boundingSphere ? geom.boundingSphere.radius * Math.max(mesh.scale.x, mesh.scale.y, mesh.scale.z) : null;
  const boxSize = geom.boundingBox ? geom.boundingBox.getSize(new THREE4.Vector3()).length() / 2 : null;
  const dim = radius ?? boxSize ?? (collider && collider.geometry?.length ? collider.geometry.length : defaultResult.dim);
  return { center, dim };
}

// src/core/aggregation.js
var Aggregation = class _Aggregation {
  /**
   * @param {string} _name
   * @param {Part[]} _parts
   * @param {Rule[]} _rules
   * @param {number|null} _rnd_seed
   */
  constructor(_name, _parts, _rules, _rnd_seed = null) {
    this.name = _name;
    this.parts = {};
    _parts.forEach((part) => {
      this.parts[part.name] = part;
    });
    this.rules = _rules;
    this.aggregated_parts = [];
    this.aggregation_queue = [];
    this.queue_values = [];
    this.queue_count = 0;
    this.rnd_seed = _rnd_seed === null ? Date.now() : _rnd_seed;
    this.randomFn = createRng(this.rnd_seed);
    this.activePartTypes = null;
    this._nextId = 0;
  }
  /**
   * Deserialize an aggregation from JSON-friendly data.
   */
  static fromData(data) {
    const d_name = data["name"];
    const normalizeList = (value) => {
      if (!value) return [];
      if (Array.isArray(value)) return value;
      if (typeof value === "object") return Object.values(value);
      return [];
    };
    const d_parts = normalizeList(data["parts"]).map((part_data) => {
      if (part_data["class_type"] === "Part") {
        return Part.fromData(part_data);
      }
      throw new Error(`Unsupported part class_type: ${part_data["class_type"]}`);
    });
    const d_rules = normalizeList(data["rules"]).map((rule) => Rule.fromData(rule));
    const d_rnd_seed = data["rnd_seed"];
    if (data["global_constraints"] && data["global_constraints"].length > 0) {
      throw new Error("Global constraints deserialization is not implemented");
    }
    if (data["catalog"]) {
      throw new Error("Part catalog deserialization is not implemented");
    }
    const aggregation = new _Aggregation(d_name, d_parts, d_rules, d_rnd_seed);
    aggregation.aggregated_parts = [];
    return aggregation;
  }
  getParts() {
    return this.aggregated_parts;
  }
  /**
   * Adjust aggregation size to match target count, mutating the visualizer when provided.
   */
  async modifyParts(newCount, waspVisualization) {
    let isCount = this.getParts().length;
    const difference = newCount - isCount;
    if (difference > 0) {
      for (let i = 0; i < difference; i++) {
        let added = this.addPartToAggregation();
        if (!added) {
          return;
        }
        const newPart = this.getParts()[this.getParts().length - 1];
        waspVisualization.addEntity(newPart);
        isCount++;
      }
    } else if (difference < 0) {
      for (let i = difference; i < 0; i++) {
        isCount--;
        let partToRemove = this.getParts()[isCount];
        waspVisualization.removeEntity(partToRemove);
        this.removePartFromAggregation(partToRemove.id);
      }
    }
  }
  /**
   * Randomly grow the aggregation by the requested number of parts.
   */
  aggregate_rnd(num) {
    let added = 0;
    let loops = 0;
    while (added < num) {
      loops += 1;
      if (loops > num * 100) {
        return `Could not place ${num - added} parts`;
      }
      if (this.aggregated_parts.length === 0) {
        if (this.addFirstPartToAggregation()) {
          added += 1;
        }
      } else {
        if (this.addPartToAggregation()) {
          added += 1;
        } else {
          return `Could not place ${num - added} parts`;
        }
      }
    }
    return true;
  }
  addFirstPartToAggregation() {
    const partKeys = this.activePartTypes ? Object.keys(this.parts).filter((k) => this.activePartTypes.has(k)) : Object.keys(this.parts);
    if (partKeys.length === 0) return false;
    let first_part = this.parts[randomChoice(this.randomFn, partKeys)];
    if (first_part !== null) {
      let first_part_trans = first_part.copy();
      first_part_trans.connections.forEach((conn) => {
        conn.generateRulesTable(this.rules);
      });
      first_part_trans.id = this._nextId++;
      this.aggregated_parts.push(first_part_trans);
      return true;
    }
    return false;
  }
  addPartToAggregation() {
    if (this.aggregated_parts.length === 0) {
      this.addFirstPartToAggregation();
      return true;
    } else {
      let next_rule = null;
      let part_01 = null;
      let conn_01_id = -1;
      let next_rule_id = -1;
      let new_rule_attempts = 0;
      while (new_rule_attempts < 1e3) {
        new_rule_attempts += 1;
        next_rule = null;
        const part_01_idx = randomInt(this.randomFn, 0, this.aggregated_parts.length - 1);
        part_01 = this.aggregated_parts[part_01_idx];
        if (part_01.active_connections.length > 0) {
          conn_01_id = randomChoice(this.randomFn, part_01.active_connections);
          let conn_01 = part_01.connections[conn_01_id];
          if (conn_01.active_rules.length > 0) {
            next_rule_id = randomChoice(this.randomFn, conn_01.active_rules);
            next_rule = conn_01.rules_table[next_rule_id];
          }
        }
        if (next_rule !== null) {
          if (this.activePartTypes && !this.activePartTypes.has(next_rule.partB)) {
            continue;
          }
          let next_part = this.parts[next_rule.partB];
          if (!next_part) {
            console.error(`Part B with id ${next_rule.partB} not found in parts.`);
            return false;
          }
          let startingPlane = next_part.connections[next_rule.connectionB].flip_pln;
          if (!startingPlane) {
            console.error(`Starting plane not found for connection B with id ${next_rule.connectionB}.`);
            return false;
          }
          let targetPlane = part_01.connections[conn_01_id].pln;
          if (!targetPlane) {
            console.error(`Target plane not found for connection 01 with id ${conn_01_id}.`);
            return false;
          }
          let orientTransform = PlaneToPlane(startingPlane, targetPlane);
          if (!orientTransform) {
            console.error(`Failed to calculate orientation transform.`);
            return false;
          }
          let next_part_trans = next_part.transform(orientTransform);
          if (!next_part_trans) {
            console.error(`Failed to transform part.`);
            return false;
          }
          let collisionDetected = false;
          const next_part_trans_collider_geo = next_part_trans.copy().collider.geometry[0];
          const next_part_trans_scale = next_part_trans.copy().geo;
          next_part_trans_scale.scale.set(0.99, 0.99, 0.99);
          for (let partToCheck of this.aggregated_parts) {
            let part1mesh = null;
            if (partToCheck.collider.geometry[0] != void 0) {
              part1mesh = partToCheck.collider.geometry[0];
            } else {
              part1mesh = partToCheck.geo;
            }
            let part2mesh = null;
            if (next_part_trans_collider_geo != void 0) {
              part2mesh = next_part_trans_collider_geo;
            } else {
              part2mesh = next_part_trans_scale;
            }
            part1mesh.updateMatrixWorld();
            part2mesh.updateMatrixWorld();
            const result = checkMeshesIntersection(part1mesh, part2mesh);
            if (result) {
              collisionDetected = true;
              break;
            }
          }
          if (collisionDetected) {
            continue;
          }
          next_part_trans.resetPart(this.rules);
          next_part_trans.active_connections = next_part_trans.active_connections.filter((conn) => conn !== next_rule.connectionB);
          next_part_trans.id = this._nextId++;
          part_01.children.push(next_part_trans.id);
          next_part_trans.parent = part_01.id;
          next_part_trans.conn_on_parent = next_rule.connectionA;
          next_part_trans.conn_to_parent = next_rule.connectionB;
          this.aggregated_parts.push(next_part_trans);
          part_01.active_connections = part_01.active_connections.filter((conn) => conn !== conn_01_id);
          return true;
        }
      }
      console.error(`No valid rule found after ${new_rule_attempts} attempts.`);
      return false;
    }
  }
  removePartFromAggregation(part_id) {
    const partIndex = this.aggregated_parts.findIndex((part) => part.id === part_id);
    if (partIndex !== -1) {
      const partToRemove = this.aggregated_parts[partIndex];
      if (partToRemove.parent !== null) {
        const parentPart = this.aggregated_parts.find((part) => part.id === partToRemove.parent);
        if (parentPart) {
          parentPart.children = parentPart.children.filter((childId) => childId !== partToRemove.id);
          parentPart.active_connections.push(partToRemove.conn_on_parent);
        }
      }
      this.aggregated_parts.splice(partIndex, 1);
      partToRemove.active_connections.forEach((conn_id) => {
        const conn = partToRemove.connections[conn_id];
        const connectedPart = this.aggregated_parts.find((part) => part.id === conn.connected_part_id);
        if (connectedPart) {
          connectedPart.active_connections = connectedPart.active_connections.filter((active_conn_id) => active_conn_id !== conn.connected_conn_id);
        }
      });
      return true;
    } else {
      console.error("Error: Part not found");
      return false;
    }
  }
  /**
   * Check if a transformed part collides with any existing aggregated part.
   * @param {Part} partTrans
   * @returns {boolean}
   */
  _checkCollision(partTrans) {
    const colliderGeo = partTrans.copy().collider.geometry[0];
    const scaledGeo = partTrans.copy().geo;
    scaledGeo.scale.set(0.99, 0.99, 0.99);
    for (const existing of this.aggregated_parts) {
      const part1mesh = existing.collider.geometry[0] != void 0 ? existing.collider.geometry[0] : existing.geo;
      const part2mesh = colliderGeo != void 0 ? colliderGeo : scaledGeo;
      part1mesh.updateMatrixWorld();
      part2mesh.updateMatrixWorld();
      if (checkMeshesIntersection(part1mesh, part2mesh)) {
        return true;
      }
    }
    return false;
  }
  /**
   * Place the first part in the aggregation by name (for manual mode).
   * @param {string} partName
   * @returns {{ success: boolean, part?: Part, error?: string }}
   */
  placeFirstPart(partName) {
    if (this.aggregated_parts.length > 0) {
      return { success: false, error: "Aggregation is not empty" };
    }
    const catalogPart = this.parts[partName];
    if (!catalogPart) {
      return { success: false, error: `Part "${partName}" not found in catalog` };
    }
    const partCopy = catalogPart.copy();
    partCopy.connections.forEach((conn) => conn.generateRulesTable(this.rules));
    partCopy.id = this._nextId++;
    this.aggregated_parts.push(partCopy);
    return { success: true, part: partCopy };
  }
  /**
   * Place a specific part at a specific connection (deterministic, for manual mode).
   * @param {number} parentPartId
   * @param {number} connectionId
   * @param {string} partName
   * @param {number} connectionBId
   * @returns {{ success: boolean, part?: Part, error?: string }}
   */
  placePartAtConnection(parentPartId, connectionId, partName, connectionBId) {
    const parentPart = this.aggregated_parts.find((p) => p.id === parentPartId);
    if (!parentPart) return { success: false, error: "Parent part not found" };
    if (!parentPart.active_connections.includes(connectionId)) {
      return { success: false, error: "Connection is not active on parent" };
    }
    const conn = parentPart.connections[connectionId];
    const catalogPart = this.parts[partName];
    if (!catalogPart) return { success: false, error: `Part "${partName}" not found in catalog` };
    const rule = conn.rules_table.find(
      (r) => r.partB === partName && r.connectionB === connectionBId
    );
    if (!rule) return { success: false, error: "No matching rule for this placement" };
    const startingPlane = catalogPart.connections[connectionBId].flip_pln;
    const targetPlane = parentPart.connections[connectionId].pln;
    const orientTransform = PlaneToPlane(startingPlane, targetPlane);
    if (!orientTransform) return { success: false, error: "Transform computation failed" };
    const partTrans = catalogPart.transform(orientTransform);
    if (!partTrans) return { success: false, error: "Part transform failed" };
    if (this._checkCollision(partTrans)) {
      return { success: false, error: "Collision detected" };
    }
    partTrans.resetPart(this.rules);
    partTrans.active_connections = partTrans.active_connections.filter((c) => c !== connectionBId);
    partTrans.id = this._nextId++;
    parentPart.children.push(partTrans.id);
    partTrans.parent = parentPart.id;
    partTrans.conn_on_parent = connectionId;
    partTrans.conn_to_parent = connectionBId;
    this.aggregated_parts.push(partTrans);
    parentPart.active_connections = parentPart.active_connections.filter((c) => c !== connectionId);
    return { success: true, part: partTrans };
  }
  /**
   * Return all open (unused) connections across aggregated parts.
   * @returns {Array<{partId: number, connectionId: number, plane: import('./plane.js').Plane, type: string, rules: Array}>}
   */
  getOpenConnections() {
    const openConns = [];
    for (const part of this.aggregated_parts) {
      for (const connIdx of part.active_connections) {
        const conn = part.connections[connIdx];
        openConns.push({
          partId: part.id,
          connectionId: connIdx,
          plane: conn.pln,
          type: conn.type,
          rules: conn.rules_table
        });
      }
    }
    return openConns;
  }
  /**
   * Compute all valid placements for a given part type.
   * Returns positions where the part can be placed without collision.
   * @param {string} partName
   * @returns {Array<{parentPartId: number, connectionId: number, partName: string, connectionBId: number, ruleIndex: number, transform: import('three').Matrix4, transformedPart: Part}>}
   */
  getValidPlacementsForPart(partName) {
    const catalogPart = this.parts[partName];
    if (!catalogPart) return [];
    const placements = [];
    const openConns = this.getOpenConnections();
    for (const openConn of openConns) {
      for (let ruleIdx = 0; ruleIdx < openConn.rules.length; ruleIdx++) {
        const rule = openConn.rules[ruleIdx];
        if (rule.partB !== partName) continue;
        const startingPlane = catalogPart.connections[rule.connectionB]?.flip_pln;
        if (!startingPlane) continue;
        const targetPlane = openConn.plane;
        const orientTransform = PlaneToPlane(startingPlane, targetPlane);
        if (!orientTransform) continue;
        const partTrans = catalogPart.transform(orientTransform);
        if (!partTrans) continue;
        if (!this._checkCollision(partTrans)) {
          placements.push({
            parentPartId: openConn.partId,
            connectionId: openConn.connectionId,
            partName,
            connectionBId: rule.connectionB,
            ruleIndex: ruleIdx,
            transform: orientTransform,
            transformedPart: partTrans
          });
        }
      }
    }
    return placements;
  }
  /**
   * Compute valid placements for a given part type, scoped to a single parent part.
   * Much cheaper than getValidPlacementsForPart when the aggregation is large.
   * @param {string} partName
   * @param {number} parentPartId
   * @returns {Array<{parentPartId: number, connectionId: number, partName: string, connectionBId: number, ruleIndex: number, transform: import('three').Matrix4, transformedPart: Part}>}
   */
  getValidPlacementsAtParent(partName, parentPartId) {
    const catalogPart = this.parts[partName];
    if (!catalogPart) return [];
    const parentPart = this.aggregated_parts.find((p) => p.id === parentPartId);
    if (!parentPart) return [];
    const placements = [];
    for (const connIdx of parentPart.active_connections) {
      const conn = parentPart.connections[connIdx];
      for (let ruleIdx = 0; ruleIdx < conn.rules_table.length; ruleIdx++) {
        const rule = conn.rules_table[ruleIdx];
        if (rule.partB !== partName) continue;
        const startingPlane = catalogPart.connections[rule.connectionB]?.flip_pln;
        if (!startingPlane) continue;
        const orientTransform = PlaneToPlane(startingPlane, conn.pln);
        if (!orientTransform) continue;
        const partTrans = catalogPart.transform(orientTransform);
        if (!partTrans) continue;
        if (!this._checkCollision(partTrans)) {
          placements.push({
            parentPartId,
            connectionId: connIdx,
            partName,
            connectionBId: rule.connectionB,
            ruleIndex: ruleIdx,
            transform: orientTransform,
            transformedPart: partTrans
          });
        }
      }
    }
    return placements;
  }
  /**
   * Compute valid placements for a given part type at a single connection.
   * Cheapest option — only checks rules for one connection slot.
   * @param {string} partName
   * @param {number} parentPartId
   * @param {number} connectionId
   * @returns {Array<{parentPartId: number, connectionId: number, partName: string, connectionBId: number, ruleIndex: number, transform: import('three').Matrix4, transformedPart: Part}>}
   */
  getValidPlacementsAtConnection(partName, parentPartId, connectionId) {
    const catalogPart = this.parts[partName];
    if (!catalogPart) return [];
    const parentPart = this.aggregated_parts.find((p) => p.id === parentPartId);
    if (!parentPart) return [];
    if (!parentPart.active_connections.includes(connectionId)) return [];
    const conn = parentPart.connections[connectionId];
    if (!conn) return [];
    const placements = [];
    for (let ruleIdx = 0; ruleIdx < conn.rules_table.length; ruleIdx++) {
      const rule = conn.rules_table[ruleIdx];
      if (rule.partB !== partName) continue;
      const startingPlane = catalogPart.connections[rule.connectionB]?.flip_pln;
      if (!startingPlane) continue;
      const orientTransform = PlaneToPlane(startingPlane, conn.pln);
      if (!orientTransform) continue;
      const partTrans = catalogPart.transform(orientTransform);
      if (!partTrans) continue;
      if (!this._checkCollision(partTrans)) {
        placements.push({
          parentPartId,
          connectionId,
          partName,
          connectionBId: rule.connectionB,
          ruleIndex: ruleIdx,
          transform: orientTransform,
          transformedPart: partTrans
        });
      }
    }
    return placements;
  }
  /**
   * Set which part types are allowed for random aggregation.
   * Pass null or undefined to allow all parts.
   * @param {string[]|null} partNames
   */
  setActivePartTypes(partNames) {
    this.activePartTypes = partNames ? new Set(partNames) : null;
  }
};
function randomChoice(rng, arr) {
  return arr[Math.floor(rng() * arr.length)];
}
function randomInt(rng, min, max) {
  return Math.floor(rng() * (max - min + 1)) + min;
}
function createRng(seed) {
  const normalizedSeed = (seed ?? Date.now()) >>> 0;
  return mulberry32(normalizedSeed);
}
function mulberry32(a) {
  return function() {
    a |= 0;
    a = a + 1831565813 | 0;
    let t = Math.imul(a ^ a >>> 15, 1 | a);
    t = t + Math.imul(t ^ t >>> 7, 61 | t) ^ t;
    return ((t ^ t >>> 14) >>> 0) / 4294967296;
  };
}

// src/core/visualizer.js
var THREE5 = __toESM(require("three"), 1);
var CameraControlsModule = __toESM(require("camera-controls"), 1);
var CameraControls = CameraControlsModule.default?.default?.install ? CameraControlsModule.default.default : CameraControlsModule.default?.install ? CameraControlsModule.default : CameraControlsModule;
CameraControls.install({ THREE: THREE5 });
function resolveElement(target) {
  if (typeof target === "string") {
    return document.querySelector(target);
  }
  return target ?? null;
}
var Visualizer = class {
  constructor(containerId = "#threejs-container", parentContainerId = containerId) {
    this.container = resolveElement(containerId);
    this.parentContainer = resolveElement(parentContainerId) || this.container;
    if (!this.container) {
      throw new Error("Visualizer container element not found.");
    }
    if (!this.parentContainer) {
      throw new Error("Visualizer parent container element not found.");
    }
    this.width = Math.max(this.parentContainer.clientWidth || this.container.clientWidth || 1, 1);
    this.height = Math.max(this.parentContainer.clientHeight || this.container.clientHeight || 1, 1);
    this.scene = new THREE5.Scene();
    this.camera = new THREE5.PerspectiveCamera(75, this.width / this.height, 0.1, 5e3);
    this.camera.up.set(0, 0, 1);
    this.camera.position.set(250, 150, 100);
    this.renderer = new THREE5.WebGLRenderer({ antialias: true, alpha: true });
    this.renderer.setClearColor(0, 0);
    this.renderer.setSize(this.width, this.height);
    this.renderer.setPixelRatio(window.devicePixelRatio);
    this.container.appendChild(this.renderer.domElement);
    this.clock = new THREE5.Clock();
    this.cameraControls = new CameraControls(this.camera, this.renderer.domElement);
    this.cameraControls.minDistance = 20;
    this.cameraControls.maxDistance = 350;
    this.cameraControls.update();
    this.scene.userData = { camera: this.camera, controls: this.cameraControls };
    this._animationFrameId = null;
    this._onResize = this.onWindowResize.bind(this);
    this.initLights();
    this.animate();
    window.addEventListener("resize", this._onResize);
  }
  initLights() {
    const directionalLight = new THREE5.DirectionalLight(16777215, 5);
    directionalLight.position.set(30, 30, -30);
    this.scene.add(directionalLight);
    const directionalLight2 = new THREE5.DirectionalLight(16777215, 1);
    directionalLight2.position.set(30, 30, 30);
    this.scene.add(directionalLight2);
    const ambientLight = new THREE5.AmbientLight(16777215, 1);
    this.scene.add(ambientLight);
  }
  addAxesHelper() {
    const gridHelper = new THREE5.GridHelper(100, 9);
    gridHelper.rotation.x = Math.PI / 2;
    const axesHelper = new THREE5.AxesHelper(10);
    axesHelper.rotation.x = Math.PI / 2;
    this.scene.add(gridHelper);
    this.scene.add(axesHelper);
  }
  animate() {
    if (!this.renderer) return;
    const delta = this.clock.getDelta();
    this.cameraControls.update(delta);
    this._animationFrameId = requestAnimationFrame(this.animate.bind(this));
    this.renderer.render(this.scene, this.camera);
  }
  onWindowResize() {
    if (!this.camera || !this.renderer || !this.parentContainer || !this.container) return;
    const newWidth = Math.max(this.parentContainer.clientWidth || this.container.clientWidth || 1, 1);
    const newHeight = Math.max(this.parentContainer.clientHeight || this.container.clientHeight || 1, 1);
    this.camera.aspect = newWidth / newHeight;
    this.camera.updateProjectionMatrix();
    this.renderer.setSize(newWidth, newHeight);
    this.renderer.setPixelRatio(window.devicePixelRatio);
    this.container.style.width = `${newWidth}px`;
    this.container.style.height = `${newHeight}px`;
  }
  addEntity(part) {
    const mesh = part.geo.clone();
    mesh.name = `${part.name}_${part.id}`;
    this.scene.add(mesh);
  }
  removeEntity(object) {
    const selectedObject = this.scene.getObjectByName(`${object.name}_${object.id}`);
    if (selectedObject) {
      this.scene.remove(selectedObject);
    }
  }
  dispose() {
    if (this._animationFrameId !== null) {
      globalThis.cancelAnimationFrame(this._animationFrameId);
      this._animationFrameId = null;
    }
    window.removeEventListener("resize", this._onResize);
    this.cameraControls?.dispose?.();
    this.renderer?.dispose?.();
  }
};

// src/core/publicApi.js
function createAggregationFromData(data) {
  return Aggregation.fromData(data);
}
function getAggregationCatalogParts(aggregation) {
  return Object.values(aggregation?.parts || {});
}
function getValidPlacementsAtParent(aggregation, partName, parentPartId) {
  if (!aggregation) return [];
  return aggregation.getValidPlacementsAtParent(partName, parentPartId);
}
function getValidPlacementsAtConnection(aggregation, partName, parentPartId, connectionId) {
  if (!aggregation) return [];
  return aggregation.getValidPlacementsAtConnection(partName, parentPartId, connectionId);
}
function getOpenConnectionsForPart(aggregation, parentPartId) {
  if (!aggregation) return [];
  const part = aggregation.aggregated_parts.find((p) => p.id === parentPartId);
  if (!part) return [];
  return part.active_connections.map((connIdx) => {
    const conn = part.connections[connIdx];
    return {
      partId: part.id,
      connectionId: connIdx,
      plane: conn.pln,
      type: conn.type,
      rules: conn.rules_table
    };
  });
}
function setActivePartTypes(aggregation, partNames) {
  if (!aggregation) return;
  aggregation.setActivePartTypes(partNames);
}

// src/types/index.js
var types = {};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Aggregation,
  Collider,
  Connection,
  Part,
  Plane,
  PlaneToPlane,
  Rule,
  Visualizer,
  checkMeshesIntersection,
  createAggregationFromData,
  generateRules,
  getAggregationCatalogParts,
  getOpenConnectionsForPart,
  getValidPlacementsAtConnection,
  getValidPlacementsAtParent,
  meshFromData,
  meshToData,
  setActivePartTypes,
  transformFromData,
  transformToData,
  types
});
